src/main/kotlin/com/angorasix/projects/presentation/infrastructure/persistence/repository/ProjectPresentationFilterRepositoryImpl.kt

```diff
@@ -21,16 +23,43 @@ class ProjectPresentationFilterRepositoryImpl(val mongoOps: ReactiveMongoOperati
     override fun findUsingFilter(filter: ListProjectPresentationsFilter): Flow<ProjectPresentation> {
         return mongoOps.find(filter.toQuery(), ProjectPresentation::class.java).asFlow()
     }
+
+    override suspend fun findByIdForContributor(
+            filter: ListProjectPresentationsFilter,
+            simpleContributor: SimpleContributor?,
+    ): ProjectPresentation? {
+        return mongoOps.find(filter.toQuery(simpleContributor), ProjectPresentation::class.java)
+                .awaitFirstOrNull()
+    }
 }
 
-private fun ListProjectPresentationsFilter.toQuery(): Query {
+private fun ListProjectPresentationsFilter.toQuery(simpleContributor: SimpleContributor? = null): Query {
     val query = Query()
+
+    val requestingOthers = adminId == null || !adminId.contains(simpleContributor?.contributorId)
+    val requestingOwn =
+            simpleContributor != null && (adminId.isNullOrEmpty() || adminId.contains(simpleContributor.contributorId))
+
+
+    ids?.let { query.addCriteria(where("_id").`in`(it)) }
+
     projectIds?.let { query.addCriteria(where("projectId").`in`(it)) }
+
     text?.let {
+        val presentationNameCriteria = where("referenceName").regex(it, "i")
         val titleCriteria = where("sections.title").regex(it, "i")
         val descriptionCriteria = where("sections.description").regex(it, "i")
-        val titleOrDescriptionCriteria = Criteria().orOperator(titleCriteria, descriptionCriteria)
-        query.addCriteria(titleOrDescriptionCriteria)
+        val titleOrDescriptionOrNameCriteria = Criteria().orOperator(titleCriteria, descriptionCriteria, presentationNameCriteria)
+        query.addCriteria(titleOrDescriptionOrNameCriteria)
+    }
+
+    if (requestingOthers) {
+        return query
     }
+
+    if (requestingOwn) {
+        query.addCriteria(where("admins").elemMatch(where("contributorId").`is`(simpleContributor?.contributorId)))
+    }
+
```

Comment:
I think we can simplify the query logic a little bit, and fix a small use case for the filtering...I think the general rule (not thinking just for our current "update" case) is: "_we should allow filtering by admins only if the requesting contributor is an actual admin of the project presentation_"

So, changes I would make (please analyze if they make sense to you):
1-  remove completely the two boolean flag variables (`requestingOthers` and `requestingOwn`)
2-  change the logic to add the `admins` filter as follows:
```
admins?.let { query.addCriteria(where("admins").`in`(admins + requestingContributor)) }
```
A few scenarios / acceptance criteria that I think this logic meets:

Having the following project presentations:
PR1 {admins: "a"}
PR2 {admins: "a", "b"}
PR3 {admins: "b"}

1- no admins filter is passed: not a problem, no filter applied

2- admins filter is used, "a" requests with filter admins=a (is our case with the update, we're passing admins=a and requestingContributor=a)

Here the service would effectively retrieve all presentations where "a" is an admin, PR1 and PR2

In our case, we're also filterin by id, so if the request was for either of these, it would work, if the request was to update PR3, then this wouldn't retrieve the Presentation due to this filter 👍 

3- admins filter is used, "a" requests with filter admins=b

Here the service would retrieve only PR2, where "a" is also an admin. Not something that we would use at this point, but I can imagine a few use cases (e.g. simply "_I ("a") want to see the projects that I administer together with  "b"_")

Let me know what you think